### PR TITLE
ci: add gcc-13, drop gcc-12

### PR DESCRIPTION
**Pull request by mrc0mmand**

### CODE REVIEW

**Review by yuwata — COMMENTED** [review]
Then, maybe better to use gcc for building bpf?

**Comment by mrc0mmand** [comment]
> Then, maybe better to use gcc for building bpf?

Yup, that's definitely the plan (as this was prompted by https://github.com/systemd/systemd/pull/27834), now I'm just checking if gcc-13 introduced some new warnings before moving further.

**Comment by medhefgo** [comment]
Maybe use the chance to drop gold completely and use mold with gcc too?

**Comment by medhefgo** [comment]
```
 ../src/boot/efi/proto/tcg.h: In function ‘tpm_log_event’:
../src/boot/efi/proto/tcg.h:57:17: note: at offset 14 into destination object ‘Event’ of size 0
   57 |         uint8_t Event[0];
```
Wups. That's supposed to be a flexible array. We could enforce strict(er) flexible arrays for sd-boot actually...

**Comment by mrc0mmand** [comment]
> ```
>  ../src/boot/efi/proto/tcg.h: In function ‘tpm_log_event’:
> ../src/boot/efi/proto/tcg.h:57:17: note: at offset 14 into destination object ‘Event’ of size 0
>    57 |         uint8_t Event[0];
> ```
> 
> Wups. That's supposed to be a flexible array. We could enforce strict(er) flexible arrays for sd-boot actually...

Interesting, I always though that `foo[0]` and `foo[]` are the [same thing](https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html) when it comes to flexible arrays. But using `Event[]` here does make the warning go away.

**Comment by medhefgo** [comment]
> Interesting, I always thought that `foo[0]` and `foo[]` are the [same thing](https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html) when it comes to flexible arrays. But using `Event[]` here does make the warning go away.

It normally is. The relatively new `-fstrict-flex-arrays(=n)` (defaults to 3) makes them distinct. While `-Wstrict-flex-arrays(=n)` (defaulting to the value from `-fstrict-flex-arrays`) turns them into warnings.

File: .github/workflows/build_test.yml [hunk 1]
```diff
@@ -25,8 +25,8 @@ jobs:
       fail-fast: false
       matrix:
         env:
-          - { COMPILER: "gcc",   COMPILER_VERSION: "11", LINKER: "bfd",  CRYPTOLIB: "gcrypt"  }
```
Comment (keszybz) [inline .github/workflows/build_test.yml:28]: Hmm, I think I would prefer to drop something else, e.g. gcc-12 or one of the other clang options. Let's keep gcc-11 working as the baseline version.

File: .github/workflows/build_test.yml [hunk 2]
```diff
@@ -26,7 +26,7 @@ jobs:
       matrix:
         env:
           - { COMPILER: "gcc",   COMPILER_VERSION: "12", LINKER: "bfd",  CRYPTOLIB: "gcrypt"  }
-          - { COMPILER: "gcc",   COMPILER_VERSION: "13", LINKER: "gold", CRYPTOLIB: "openssl" }
+          - { COMPILER: "gcc",   COMPILER_VERSION: "13", LINKER: "mold", CRYPTOLIB: "openssl" }
```
Comment (keszybz) [inline .github/workflows/build_test.yml]: Same here. Unless there's some other justification, let's keep `gold` in at least one path.

**Review by keszybz — APPROVED** [review]

**Review by bluca — APPROVED** [review]

**Pull request merged**